server/src/transactions/use-cases/registerTransaction/register-transaction.ts

```diff
@@ -1,14 +1,33 @@
-import { Injectable } from '@nestjs/common';
+import { Injectable, UnauthorizedException } from '@nestjs/common';
 
-import { ITransactionsRepository } from '@transactions/interfaces';
+import { Reasons, Roles } from '@shared/constants';
 import { RegisterTransactionDTO } from '@transactions/dto';
 import { Transaction } from '@transactions/infra/typeorm/entities/transaction.entity';
+import { ITransactionsRepository } from '@transactions/interfaces';
+import { IUsersRepository } from '@users/interfaces/IUsersRepository';
 
 @Injectable()
 export class RegisterTransaction {
-  constructor(private transactionsRepository: ITransactionsRepository) {}
+  constructor(
+    private transactionsRepository: ITransactionsRepository,
+    private usersRepository: IUsersRepository,
+  ) {}
 
   async execute(data: RegisterTransactionDTO): Promise<Transaction> {
+    const responsible = await this.usersRepository.findOne(data.responsible);
+
+    if (
```

Comment:
Junta esses IFs em um, pois eles retornam o mesmo erro.

if (
  (responsible.role === Roles.COLLABORATOR && data.reason !== Reasons.REDEEM) ||
  (responsible.role === Roles.ACADEMY && data.reason !== Reasons.COLLABORATION)
) {
  throw new UnauthorizedException('You do not have permission');
}

Além disso, melhora essa validação gigantesca deixando enxuta e tacando um includes